File: lib/product-feed.js
// Single source of truth for everything derived from a Supabase product row:
// its public URL, its price, its images, and the three published
// representations of it - the Google Merchant Center feed, the Product JSON-LD
// embedded in the product page, and the sitemap.
//
// The point of keeping these together is that a product cannot say $16.00 on
// the page, 16.00 in JSON-LD and something else in the feed. Google treats that
// mismatch as a policy violation, so every representation below reads the same
// fields from the same row.

const SITE_ORIGIN = 'https://dwellingdream.shop';

// These palettes are Dwelling Dream's own work. The paint manufacturer named on
// a palette (Behr, Sherwin Williams, Benjamin Moore) is the paint the colors
// refer to - it is NOT the brand of the product being sold, and claiming it
// would misrepresent these as official manufacturer products. The manufacturer
// goes in product_type instead.
const BRAND = 'Dwelling Dream';

// Google's own taxonomy (https://support.google.com/merchants/answer/6324436).
// Deliberately left unset: google_product_category is optional, Google
// auto-classifies when it is absent, and a confidently-wrong category is worse
// than none. Fill in an internal category here to override, e.g.
//   'Behr': 'Home & Garden > Decor'
// Verbatim path from Google's published taxonomy (id 500044,
// taxonomy-with-ids.en-US.txt, version 2021-09-21). Must stay verbatim - an
// unrecognised path is a hard error, not a warning.
//
// Three values have been tried, and the history matters before changing it:
//   unset                      -> warning 157, "missing, may limit visibility"
//   Home & Garden > Decor      -> warning 126, only two levels deep
//   Media > Books > E-books    -> ERROR, "Digital books not supported and
//                                 cannot be listed on Shopping". Google bans
//                                 ebooks outright, so that value did not just
//                                 warn, it disqualified every item.
// Posters/Prints is four levels, is not a banned category, and is defensible
// for printable colour palettes with room photography.
// https://support.google.com/merchants/answer/6324436
const DEFAULT_GOOGLE_PRODUCT_CATEGORY = 'Home & Garden > Decor > Artwork > Posters, Prints, & Visual Artwork';
const GOOGLE_PRODUCT_CATEGORY = {
  'Behr': DEFAULT_GOOGLE_PRODUCT_CATEGORY,
  'Sherwin Williams': DEFAULT_GOOGLE_PRODUCT_CATEGORY,
  'Benjamin Moore': DEFAULT_GOOGLE_PRODUCT_CATEGORY
};

// Google and Pinterest fetch every image in the feed in a burst. Pointed
// straight at Supabase Storage that returns HTTP 429 for a large share of them
// (reproduced: 176 of 430 requests failed), because the free tier rate-limits.
// Serving them from this site instead puts Hostinger's CDN in front, so the
// crawler hits a cached edge and Supabase is asked for each object once.
const SUPABASE_PUBLIC_IMAGE = /^https:\/\/[a-z0-9-]+\.supabase\.co\/storage\/v1\/object\/public\/product-images\/(.+)$/;

function publicImageUrl(url) {
  const match = SUPABASE_PUBLIC_IMAGE.exec(String(url || ''));
  return match ? `${SITE_ORIGIN}/product-image/${match[1]}` : url;
}

function slugify(value) {
  return String(value || '').toLowerCase().trim().replace(/[^a-z0-9]+/g, '-').replace(/^-+|-+$/g, '');
}

// The original rule: derive the slug from category + title. Kept because every
// URL published before products.slug existed was built this way, and those
// links must keep resolving forever.
function derivedSlug(product) {
  const title = product.title || '';
  const category = product.category || '';
  const base = title.toLowerCase().includes(category.toLowerCase()) ? title : `${category} ${title}`;
  return slugify(base);
}

// The canonical slug. Once products.slug is populated (see
// supabase/schema_product_slug.sql) a product's URL is fixed at creation and
// survives any retitle. Falls back to the derived rule when the column is
// absent or empty, so this is safe to run before that migration.
function productSlug(product) {
  const stored = typeof product.slug === 'string' ? product.slug.trim() : '';
  return stored || derivedSlug(product);
}

// A product answers to its canonical slug AND to its derived one, so a URL
// shared before a rename still lands on the right product instead of 404ing.
function matchesSlug(product, slug) {
  if (!slug) return false;
  return slug === productSlug(product) || slug === derivedSlug(product);
}

// The currency is carried in the URL rather than guessed from the visitor's
// IP. Google crawls the GB feed's links from the UK and must see the GBP
// price; a query parameter guarantees that, geo-detection does not.
function productUrl(product, code) {
  const base = `${SITE_ORIGIN}/Dwelling%20Dream%20Product.dc.html?slug=${productSlug(product)}`;
  const cfg = currencyConfig(code);
  return cfg.code === DEFAULT_CURRENCY ? base : `${base}&currency=${cfg.code}`;
}

// Note: this maps to site-hosted URLs and is used by the feeds and JSON-LD
// only. /api/products deliberately keeps returning the raw Supabase URLs,
// because the admin panel round-trips those exact strings back as keepImages
// when editing a product.
function productImages(product) {
  return (product.images || [])
    .filter(url => typeof url === 'string' && url.startsWith('https://'))
    .map(publicImageUrl);
}

function priceAmount(product) {
  const value = Number(product.price);
  return Number.isFinite(value) && value > 0 ? value.toFixed(2) : null;
}

function currencyOf(product) {
  return String(product.currency || 'USD').toUpperCase();
}

// `active` is the only publication state this schema has - there is no draft,
// sold-out or hidden flag. Digital downloads never run out of stock, so an
// active product is always in_stock.
function availabilityOf(product) {
  return product.active === false ? 'out_of_stock' : 'in_stock';
}

// `code` is the storefront currency the listing is for. A product with no
// price in that currency is not listable there - it is dropped from that
// country's feed rather than advertised at a price checkout would not honour.
function isListable(product, code = DEFAULT_CURRENCY) {
  return Boolean(
    product &&
    product.active !== false &&
    product.id &&
    (product.title || '').trim() &&
    (product.description || '').trim() &&
    priceIn(product, code) &&
    productImages(product).length
  );
}

function escapeXml(value) {
  return String(value == null ? '' : value)
    .replace(/&/g, '&amp;')
    .replace(/</g, '&lt;')
    .replace(/>/g, '&gt;')
    .replace(/"/g, '&quot;')
    .replace(/'/g, '&apos;');
}

// Google rejects raw control characters even when they are XML-escaped.
function cleanText(value, maxLength) {
  const text = String(value == null ? '' : value)
    .replace(/[\u0000-\u001F\u007F]/g, ' ')
    .replace(/\s+/g, ' ')
    .trim();
  return maxLength && text.length > maxLength ? text.slice(0, maxLength - 1).trimEnd() + '\u2026' : text;
}

function feedItem(product, code = DEFAULT_CURRENCY) {
  const images = productImages(product);
  const cfg = currencyConfig(code);
  const item = {
    id: String(product.id),
    title: cleanText(product.title, 150),
    description: cleanText(product.description, 5000),
    link: productUrl(product, cfg.code),
    image_link: images[0],
    // Google and Pinterest both accept up to 10 additional images. This was
    // temporarily cut to 4 while product images were served straight from
    // Supabase Storage, whose free tier rate-limited the crawl (Google error
    // 1205 / warning 1222, HTTP 429). Images now go through /product-image/
    // behind Hostinger's CDN, so the full set is affordable again.
    additional_image_link: images.slice(1, 11),
    availability: availabilityOf(product),
    price: `${priceIn(product, cfg.code)} ${cfg.code}`,
    currency: cfg.code,
    shippingCountry: cfg.country,
    brand: BRAND,
    condition: 'new',
    // No GTIN/UPC/EAN exists for these products and inventing one is a policy
    // violation, so tell Google none exists and identify by brand + mpn.
    identifier_exists: 'no',
    mpn: product.sku || undefined,
    product_type: product.category ? `Paint Color Palettes > ${product.category}` : 'Paint Color Palettes',
    // Digital downloads are never shipped.
    shipping_weight: undefined,
    google_product_category: GOOGLE_PRODUCT_CATEGORY[product.category] || undefined
  };
  return item;
}

// Pinterest ingests the same RSS + g: namespace shape as Google but spells
// several values differently. Taken from Pinterest's own sample feed:
//   availability      "In Stock"      (Google: in_stock)
//   condition         "New"           (Google: new)
//   identifier_exists "FALSE"         (Google: no)
// Sending Google's spellings to Pinterest gets items rejected, so the two
// dialects are generated from the same item data but written out differently.
const PINTEREST_AVAILABILITY = {
  in_stock: 'In Stock',
  out_of_stock: 'Out of Stock'
};

// Nothing is ever shipped - these are downloads - so a zero shipping cost is
// truthful. Google reported "products are set to show in countries that lack
// shipping information", which it raises even for digital goods, so both feeds
// now carry an explicit free-shipping line per targeted country.
//
// This list must match the countries targeted in Merchant Center. Adding a
// country here does NOT make the price valid there: Google requires the price
// currency to match the target country, and this store prices only in USD.
const SHIPPING_COUNTRIES = ['US', 'GB'];

// Supported storefront currencies. The site prices in USD by default; the UK
// is offered a separate fixed GBP price rather than a converted one, because
// Google requires the feed price to equal the landing-page price and a live
// rate cannot guarantee that (see supabase/schema_price_gbp.sql).
//
// `field` is the product property holding the amount, `country` is the
// Merchant Center target the matching feed is built for.
const CURRENCIES = {
  USD: { code: 'USD', symbol: '$', field: 'price', country: 'US' },
  GBP: { code: 'GBP', symbol: '£', field: 'priceGbp', country: 'GB' }
};
const DEFAULT_CURRENCY = 'USD';

function currencyConfig(code) {
  return CURRENCIES[String(code || '').toUpperCase()] || CURRENCIES[DEFAULT_CURRENCY];
}

// Amount for a given currency, or null when this product has no price in it.
// Null means "exclude from that country's feed" - never "fall back to USD",
// which would advertise a price the checkout would not honour.
function priceIn(product, code) {
  const cfg = currencyConfig(code);
  const value = Number(product[cfg.field]);
  return Number.isFinite(value) && value > 0 ? value.toFixed(2) : null;
}
const PINTEREST_SHIPPING_COUNTRY = 'US';

// g:tax tells Pinterest how much tax to ADD on top of the advertised price.
// The store prices tax-inclusive - $16.00 is what the buyer pays - so the
// amount added at checkout is genuinely zero. Rate 0 here means "nothing is
// added to the shown price", NOT "this product is exempt from tax"; it says
// nothing about what the seller owes and remits.
//
// If the store ever switches to adding tax at checkout, this must change to
// the real rate at the same time as the checkout does, or Pinterest will
// advertise a price lower than buyers are charged.
const PINTEREST_TAX = {
  country: 'US',
  rate: 0,
  taxShip: 'n'
};

function buildFeedXml(products, options = {}) {
  const { dialect = 'google', generatedAt = new Date(), currency = DEFAULT_CURRENCY } = options;
  const pinterest = dialect === 'pinterest';
  const cfg = currencyConfig(currency);
  const items = products.filter(p => isListable(p, cfg.code)).map(p => feedItem(p, cfg.code));

  const body = items.map(item => {
    const lines = [];
    const push = (tag, value) => lines.push(`      <${tag}>${escapeXml(value)}</${tag}>`);

    push('g:id', item.id);
    push('title', item.title);
    push('description', item.description);
    push('g:product_type', item.product_type);
    if (item.google_product_category) push('g:google_product_category', item.google_product_category);
    push('link', item.link);
    push('g:image_link', item.image_link);
    push('g:condition', pinterest ? 'New' : item.condition);
    push('g:availability', pinterest ? (PINTEREST_AVAILABILITY[item.availability] || item.availability) : item.availability);
    push('g:price', item.price);
    if (item.mpn) push('g:mpn', item.mpn);
    push('g:brand', item.brand);

    if (!pinterest) {
      // One country per feed: a GBP feed ships to GB, a USD feed to US.
      // Mixing them would advertise a price in the wrong currency for a
      // country, which Google rejects.
      lines.push('      <g:shipping>');
      lines.push(`          <g:country>${escapeXml(item.shippingCountry)}</g:country>`);
      lines.push(`          <g:price>0 ${escapeXml(item.currency)}</g:price>`);
      lines.push('      </g:shipping>');
    }

    if (pinterest) {
      lines.push('      <g:tax>');
      lines.push(`          <g:country>${escapeXml(PINTEREST_TAX.country)}</g:country>`);
      lines.push(`          <g:rate>${escapeXml(PINTEREST_TAX.rate)}</g:rate>`);
      lines.push(`          <g:tax_ship>${escapeXml(PINTEREST_TAX.taxShip)}</g:tax_ship>`);
      lines.push('      </g:tax>');
      lines.push('      <g:shipping>');
      lines.push(`          <g:country>${escapeXml(PINTEREST_SHIPPING_COUNTRY)}</g:country>`);
      lines.push(`          <g:price>0 ${escapeXml(item.currency)}</g:price>`);
      lines.push('      </g:shipping>');
    }

    push('g:identifier_exists', pinterest ? 'FALSE' : item.identifier_exists);
    item.additional_image_link.forEach(url => push('g:additional_image_link', url));

    return `    <item>\n${lines.join('\n')}\n    </item>`;
  }).join('\n');

  return `<?xml version="1.0" encoding="UTF-8"?>
<rss version="2.0" xmlns:g="http://base.google.com/ns/1.0">
  <channel>
    <title>Dwelling Dream</title>
    <link>${escapeXml(SITE_ORIGIN)}</link>
    <description>Digital paint color palettes, guides and planners - instant download.</description>
    <lastBuildDate>${escapeXml(generatedAt.toUTCString())}</lastBuildDate>
${body}
  </channel>
</rss>
`;
}

// Mirrors feedItem() field for field so the page and the feed can never
// disagree about price, availability, image or URL.
// A product may have no price in the requested currency - a newly added one
// has no GBP price until an admin sets it. The page falls back to the default
// currency rather than rendering a null price. The feeds do NOT fall back:
// they drop the item instead, so nothing is ever advertised in a currency the
// checkout would refuse.
function effectiveCurrency(product, code) {
  const cfg = currencyConfig(code);
  return priceIn(product, cfg.code) ? cfg.code : DEFAULT_CURRENCY;
}

function productJsonLd(product, code = DEFAULT_CURRENCY) {
  const images = productImages(product);
  const cfg = currencyConfig(effectiveCurrency(product, code));
  const schema = {
    '@context': 'https://schema.org',
    '@type': 'Product',
    name: cleanText(product.title, 150),
    description: cleanText(product.description, 5000),
    image: images.slice(0, 10),
    url: productUrl(product, cfg.code),
    brand: { '@type': 'Brand', name: BRAND },
    offers: {
      '@type': 'Offer',
      price: priceIn(product, cfg.code),
      priceCurrency: cfg.code,
      availability: availabilityOf(product) === 'in_stock'
        ? 'https://schema.org/InStock'
        : 'https://schema.org/OutOfStock',
      url: productUrl(product, cfg.code),
      itemCondition: 'https://schema.org/NewCondition'
    }
  };
  if (product.sku) schema.sku = product.sku;
  if (product.category) schema.category = `Paint Color Palettes > ${product.category}`;
  return schema;
}

// Developer-readable feed validation. Returns the problems that would make
// Google reject or silently drop an item, so they can be caught before a
// scheduled fetch does.
function validateProducts(products) {
  const errors = [];
  const warnings = [];
  const seenIds = new Map();
  const seenUrls = new Map();

  for (const product of products) {
    const label = product.sku || product.id || '(unknown)';

    if (!product.id) errors.push(`${label}: missing id`);
    else if (seenIds.has(String(product.id))) errors.push(`${label}: duplicate id "${product.id}" (also ${seenIds.get(String(product.id))})`);
    else seenIds.set(String(product.id), label);

    if (!(product.title || '').trim()) errors.push(`${label}: missing title`);
    if (!(product.description || '').trim()) errors.push(`${label}: missing description`);
    if (!priceAmount(product)) errors.push(`${label}: missing or invalid price (${JSON.stringify(product.price)})`);
    if (!/^[A-Z]{3}$/.test(currencyOf(product))) errors.push(`${label}: invalid currency "${product.currency}"`);

    const url = productUrl(product);
    if (!url.startsWith('https://')) errors.push(`${label}: product URL is not HTTPS`);
    if (seenUrls.has(url)) errors.push(`${label}: duplicate product URL with ${seenUrls.get(url)} - slugs collide`);
    else seenUrls.set(url, label);

    const allImages = product.images || [];
    const httpsImages = productImages(product);
    if (!allImages.length) errors.push(`${label}: no images`);
    else if (!httpsImages.length) errors.push(`${label}: no HTTPS images (${allImages.length} non-HTTPS dropped)`);
    else if (httpsImages.length < allImages.length) warnings.push(`${label}: ${allImages.length - httpsImages.length} non-HTTPS image(s) dropped`);

    if (!product.sku) warnings.push(`${label}: no SKU - item will be sent without g:mpn`);
    if (!GOOGLE_PRODUCT_CATEGORY[product.category]) warnings.push(`${label}: no google_product_category mapped for "${product.category}" - Google will auto-classify`);
    if (product.active === false) warnings.push(`${label}: inactive - excluded from feed`);
  }

  return { errors, warnings };
}

module.exports = {
  SITE_ORIGIN,
  BRAND,
  GOOGLE_PRODUCT_CATEGORY,
  slugify,
  derivedSlug,
  productSlug,
  matchesSlug,
  productUrl,
  productImages,
  publicImageUrl,
  priceAmount,
  priceIn,
  currencyConfig,
  effectiveCurrency,
  CURRENCIES,
  DEFAULT_CURRENCY,
  currencyOf,
  availabilityOf,
  isListable,
  escapeXml,
  cleanText,
  feedItem,
  buildFeedXml,
  productJsonLd,
  validateProducts
};

File: lib/product_feed.py
"""Python mirror of lib/product-feed.js.

server.js is the deployed runtime; server.py is kept as a parallel
implementation, so this module exists to stop the two diverging. Any change to
one must be made to the other - the field mapping, the brand, the category map
and the availability rule all have to agree, because both can render the same
product page and the same Merchant Center feed.

Single source of truth for everything derived from a Supabase product row: its
public URL, price, images, and the three published representations - the Google
Merchant Center feed, the Product JSON-LD on the product page, and the sitemap.
"""

import json
import re

SITE_ORIGIN = "https://dwellingdream.shop"

# These palettes are Dwelling Dream's own work. The paint manufacturer named on
# a palette (Behr, Sherwin Williams, Benjamin Moore) is the paint the colors
# refer to - it is NOT the brand of the product being sold, and claiming it
# would misrepresent these as official manufacturer products. The manufacturer
# goes in product_type instead.
BRAND = "Dwelling Dream"

# Google's taxonomy (https://support.google.com/merchants/answer/6324436).
# Deliberately empty: google_product_category is optional, Google auto-classifies
# when absent, and a confidently-wrong category is worse than none. Add an
# internal category here to override, e.g. {"Behr": "Home & Garden > Decor"}.
# Google warned that missing values limit visibility in recommendations and
# shopping surfaces. Google's taxonomy has no entry for "digital paint palette
# guide"; Decor is the closest honest fit for a home-colour lookbook. Must be a
# verbatim path from https://support.google.com/merchants/answer/6324436
DEFAULT_GOOGLE_PRODUCT_CATEGORY = "Home & Garden > Decor"
GOOGLE_PRODUCT_CATEGORY = {
    "Behr": DEFAULT_GOOGLE_PRODUCT_CATEGORY,
    "Sherwin Williams": DEFAULT_GOOGLE_PRODUCT_CATEGORY,
    "Benjamin Moore": DEFAULT_GOOGLE_PRODUCT_CATEGORY,
}

# Google and Pinterest fetch every image in a burst. Pointed straight at
# Supabase Storage that returns HTTP 429 for a large share of them (reproduced:
# 176 of 430), because the free tier rate-limits. Serving them from this site
# puts Hostinger's CDN in front, so Supabase is asked for each object once.
SUPABASE_PUBLIC_IMAGE = re.compile(
    r"^https://[a-z0-9-]+\.supabase\.co/storage/v1/object/public/product-images/(.+)$")


def public_image_url(url):
    m = SUPABASE_PUBLIC_IMAGE.match(str(url or ""))
    return f"{SITE_ORIGIN}/product-image/{m.group(1)}" if m else url

_CONTROL_CHARS = re.compile(r"[\x00-\x1F\x7F]")


def slugify(value):
    text = re.sub(r"[^a-z0-9]+", "-", str(value or "").lower().strip())
    return text.strip("-")


def derived_slug(product):
    """The original rule: derive from category + title. Kept because every URL
    published before products.slug existed was built this way, and those links
    must keep resolving forever."""
    title = product.get("title") or ""
    category = product.get("category") or ""
    base = title if category.lower() in title.lower() else f"{category} {title}"
    return slugify(base)


def product_slug(product):
    """The canonical slug. Once products.slug is populated (see
    supabase/schema_product_slug.sql) a product's URL is fixed at creation and
    survives any retitle. Falls back to the derived rule when the column is
    absent, so this is safe before that migration."""
    stored = product.get("slug")
    stored = stored.strip() if isinstance(stored, str) else ""
    return stored or derived_slug(product)


def matches_slug(product, slug):
    """A product answers to its canonical slug AND its derived one, so a URL
    shared before a rename still lands on the right product instead of 404ing."""
    if not slug:
        return False
    return slug in (product_slug(product), derived_slug(product))


def product_url(product):
    return f"{SITE_ORIGIN}/Dwelling%20Dream%20Product.dc.html?slug={product_slug(product)}"


def product_images(product):
    """Maps to site-hosted URLs; used by the feeds and JSON-LD only. The
    products API deliberately keeps returning raw Supabase URLs, because the
    admin panel round-trips those exact strings back as keepImages."""
    return [public_image_url(u) for u in (product.get("images") or [])
            if isinstance(u, str) and u.startswith("https://")]


def price_amount(product):
    try:
        value = float(product.get("price"))
    except (TypeError, ValueError):
        return None
    return f"{value:.2f}" if value > 0 else None


def currency_of(product):
    return str(product.get("currency") or "USD").upper()


def availability_of(product):
    # `active` is the only publication state this schema has - no draft,
    # sold-out or hidden flag. Digital downloads never run out of stock.
    return "out_of_stock" if product.get("active") is False else "in_stock"


def is_listable(product):
    return bool(
        product
        and product.get("active") is not False
        and product.get("id")
        and (product.get("title") or "").strip()
        and (product.get("description") or "").strip()
        and price_amount(product)
        and product_images(product)
    )


def escape_xml(value):
    return (
        str("" if value is None else value)
        .replace("&", "&amp;")
        .replace("<", "&lt;")
        .replace(">", "&gt;")
        .replace('"', "&quot;")
        .replace("'", "&apos;")
    )


def clean_text(value, max_length=None):
    """Google rejects raw control characters even when XML-escaped."""
    text = _CONTROL_CHARS.sub(" ", str("" if value is None else value))
    text = re.sub(r"\s+", " ", text).strip()
    if max_length and len(text) > max_length:
        return text[: max_length - 1].rstrip() + "…"
    return text


def feed_item(product):
    images = product_images(product)
    category = product.get("category")
    return {
        "id": str(product.get("id")),
        "title": clean_text(product.get("title"), 150),
        "description": clean_text(product.get("description"), 5000),
        "link": product_url(product),
        "image_link": images[0] if images else None,
        "additional_image_link": images[1:5],
        "availability": availability_of(product),
        "price": f"{price_amount(product)} {currency_of(product)}",
        "brand": BRAND,
        "condition": "new",
        # No GTIN/UPC/EAN exists for these and inventing one is a policy
        # violation, so tell Google none exists and identify by brand + mpn.
        "identifier_exists": "no",
        "mpn": product.get("sku") or None,
        "product_type": f"Paint Color Palettes > {category}" if category else "Paint Color Palettes",
        "google_product_category": GOOGLE_PRODUCT_CATEGORY.get(category),
    }


# Pinterest ingests the same RSS + g: namespace shape as Google, but spells
# availability with a space ("in stock") rather than an underscore.
PINTEREST_AVAILABILITY = {"in_stock": "in stock", "out_of_stock": "out of stock"}


def build_feed_xml(products, generated_at=None, dialect="google"):
    from email.utils import format_datetime
    from datetime import datetime, timezone

    generated_at = generated_at or datetime.now(timezone.utc)
    blocks = []
    for product in products:
        if not is_listable(product):
            continue
        item = feed_item(product)
        pinterest = dialect == "pinterest"
        availability = PINTEREST_AVAILABILITY.get(item["availability"], item["availability"]) if pinterest else item["availability"]
        lines = [
            f"      <g:id>{escape_xml(item['id'])}</g:id>",
            f"      <title>{escape_xml(item['title'])}</title>",
            f"      <description>{escape_xml(item['description'])}</description>",
            f"      <link>{escape_xml(item['link'])}</link>",
            f"      <g:image_link>{escape_xml(item['image_link'])}</g:image_link>",
        ]
        lines += [f"      <g:additional_image_link>{escape_xml(u)}</g:additional_image_link>"
                  for u in item["additional_image_link"]]
        lines += [
            f"      <g:availability>{escape_xml(availability)}</g:availability>",
            f"      <g:price>{escape_xml(item['price'])}</g:price>",
            f"      <g:brand>{escape_xml(item['brand'])}</g:brand>",
            f"      <g:condition>{escape_xml(item['condition'])}</g:condition>",
        ]
        if not pinterest:
            lines.append(f"      <g:identifier_exists>{escape_xml(item['identifier_exists'])}</g:identifier_exists>")
        if item["mpn"]:
            lines.append(f"      <g:mpn>{escape_xml(item['mpn'])}</g:mpn>")
        if item["google_product_category"]:
            lines.append(f"      <g:google_product_category>{escape_xml(item['google_product_category'])}</g:google_product_category>")
        lines.append(f"      <g:product_type>{escape_xml(item['product_type'])}</g:product_type>")
        blocks.append("    <item>\n" + "\n".join(lines) + "\n    </item>")

    return (
        '<?xml version="1.0" encoding="UTF-8"?>\n'
        '<rss version="2.0" xmlns:g="http://base.google.com/ns/1.0">\n'
        "  <channel>\n"
        "    <title>Dwelling Dream</title>\n"
        f"    <link>{escape_xml(SITE_ORIGIN)}</link>\n"
        "    <description>Digital paint color palettes, guides and planners - instant download.</description>\n"
        f"    <lastBuildDate>{escape_xml(format_datetime(generated_at))}</lastBuildDate>\n"
        + "\n".join(blocks)
        + "\n  </channel>\n</rss>\n"
    )


def product_json_ld(product):
    """Mirrors feed_item() field for field so page and feed cannot disagree."""
    images = product_images(product)
    schema = {
        "@context": "https://schema.org",
        "@type": "Product",
        "name": clean_text(product.get("title"), 150),
        "description": clean_text(product.get("description"), 5000),
        "image": images[:10],
        "url": product_url(product),
        "brand": {"@type": "Brand", "name": BRAND},
        "offers": {
            "@type": "Offer",
            "price": price_amount(product),
            "priceCurrency": currency_of(product),
            "availability": "https://schema.org/InStock"
            if availability_of(product) == "in_stock"
            else "https://schema.org/OutOfStock",
            "url": product_url(product),
            "itemCondition": "https://schema.org/NewCondition",
        },
    }
    if product.get("sku"):
        schema["sku"] = product["sku"]
    if product.get("category"):
        schema["category"] = f"Paint Color Palettes > {product['category']}"
    return schema


def product_json_ld_script(product):
    """`</` is escaped so a description containing "</script>" cannot break out."""
    payload = json.dumps(product_json_ld(product), ensure_ascii=False).replace("<", "\\u003c")
    return f'<script type="application/ld+json">{payload}</script>\n'


def validate_products(products):
    errors, warnings = [], []
    seen_ids, seen_urls = {}, {}

    for product in products:
        label = product.get("sku") or product.get("id") or "(unknown)"

        if not product.get("id"):
            errors.append(f"{label}: missing id")
        elif str(product["id"]) in seen_ids:
            errors.append(f"{label}: duplicate id \"{product['id']}\" (also {seen_ids[str(product['id'])]})")
        else:
            seen_ids[str(product["id"])] = label

        if not (product.get("title") or "").strip():
            errors.append(f"{label}: missing title")
        if not (product.get("description") or "").strip():
            errors.append(f"{label}: missing description")
        if not price_amount(product):
            errors.append(f"{label}: missing or invalid price ({product.get('price')!r})")
        if not re.match(r"^[A-Z]{3}$", currency_of(product)):
            errors.append(f"{label}: invalid currency \"{product.get('currency')}\"")

        url = product_url(product)
        if not url.startswith("https://"):
            errors.append(f"{label}: product URL is not HTTPS")
        if url in seen_urls:
            errors.append(f"{label}: duplicate product URL with {seen_urls[url]} - slugs collide")
        else:
            seen_urls[url] = label

        all_images = product.get("images") or []
        https_images = product_images(product)
        if not all_images:
            errors.append(f"{label}: no images")
        elif not https_images:
            errors.append(f"{label}: no HTTPS images ({len(all_images)} non-HTTPS dropped)")
        elif len(https_images) < len(all_images):
            warnings.append(f"{label}: {len(all_images) - len(https_images)} non-HTTPS image(s) dropped")

        if not product.get("sku"):
            warnings.append(f"{label}: no SKU - item will be sent without g:mpn")
        if not GOOGLE_PRODUCT_CATEGORY.get(product.get("category")):
            warnings.append(f"{label}: no google_product_category mapped for \"{product.get('category')}\" - Google will auto-classify")
        if product.get("active") is False:
            warnings.append(f"{label}: inactive - excluded from feed")

    return errors, warnings


def build_sitemap_xml(products):
    static_paths = [
        "/",
        "/Dwelling%20Dream%20Palettes.dc.html",
        "/Dwelling%20Dream%20About.dc.html",
        "/Dwelling%20Dream%20Help.dc.html",
    ]
    urls = [f"  <url>\n    <loc>{escape_xml(SITE_ORIGIN + p)}</loc>\n  </url>" for p in static_paths]
    for product in products:
        if not is_listable(product):
            continue
        lastmod = str(product.get("updatedAt") or "")[:10]
        entry = f"  <url>\n    <loc>{escape_xml(product_url(product))}</loc>"
        if lastmod:
            entry += f"\n    <lastmod>{escape_xml(lastmod)}</lastmod>"
        urls.append(entry + "\n  </url>")
    return (
        '<?xml version="1.0" encoding="UTF-8"?>\n'
        '<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">\n'
        + "\n".join(urls)
        + "\n</urlset>\n"
    )


ROBOTS_TXT = "\n".join([
    "User-agent: *",
    "Allow: /",
    "Disallow: /admin.html",
    "Disallow: /login.html",
    "Disallow: /api/",
    # Longest-match wins, so this re-permits the Merchant Center feed that the
    # broader /api/ rule would otherwise cover.
    "Allow: /api/google-shopping-feed",
    "Allow: /api/pinterest-feed",
    "",
    f"Sitemap: {SITE_ORIGIN}/sitemap.xml",
    "",
])
